core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/config/UserStoreConfigXMLProcessor.java

```diff
@@ -430,143 +427,139 @@ private static PrivateKey getPrivateKey() {
-                log.debug("Deserialization failed since cipher string is not representing cipher with metadata");
-            }
-            return null;
         }
-    }
 
-    /**
-     * Holds encrypted cipher with related metadata.
-     *
-     * IMPORTANT: this is copy of org.wso2.carbon.core.util.CipherHolder, what ever changes applied here need to update
-     *              on above
-     */
-    private class CipherHolder {
-
-        // Base64 encoded ciphertext.
-        private String c;
+    }
 
-        // Transformation used for encryption, default is "RSA".
-        private String t = "RSA";
+    private String decryptWithPrimaryKeyStore(String propValue) throws CryptoException {
 
-        // Thumbprint of the certificate.
-        private String tp;
+        Cipher keyStoreCipher = null;
+        String cipherTransformation = System.getProperty(CIPHER_TRANSFORMATION_SYSTEM_PROPERTY);
+        byte[] cipherTextBytes = Base64.decode(propValue.trim());
+        byte[] plainTextBytes = new byte[0];
 
-        // Digest used to generate certificate thumbprint.
-        private String tpd;
+        privateKey = (privateKey == null) ? getPrivateKey() : privateKey;
+        if (privateKey == null) {
+            throw new CryptoException(
+                    "Private key initialization failed. Cannot decrypt the userstore password.");
+        }
 
+        try {
+            if (cipherTransformation != null) {
+                // extract the original cipher if custom transformation is used configured in carbon.properties.
+                CipherMetaDataHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
+                if (cipherHolder != null) {
+                    // cipher with meta data.
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cipher transformation for decryption : " + cipherHolder.getTransformation());
+                    }
 
-        public String getTransformation() {
-            return t;
-        }
+                    keyStoreCipher = Cipher.getInstance(cipherHolder.getTransformation(), "BC");
 
-        public void setTransformation(String transformation) {
-            this.t = transformation;
-        }
+                    cipherTextBytes = cipherHolder.getCipherBase64Decoded();
+                } else {
+                    // If the ciphertext is not a self-contained, directly decrypt using transformation configured in
+                    // carbon.properties file
 
-        public String getCipherText() {
-            return c;
-        }
+                    keyStoreCipher = Cipher.getInstance(cipherTransformation, "BC");
 
-        public byte[] getCipherBase64Decoded() {
-            return Base64.decode(c);
-        }
+                }
+            } else {
+                // If reach here, user have removed org.wso2.CipherTransformation property or carbon.properties file
+                // hence RSA is considered as default transformation
```

Comment:
Shall we provide a debug log as well with this information?

Reply:
fixed with 8d1a059d01c4d955897f1834cc148dff8df0c730